### ISSUE TITLE
Non-performing model/model weights uniformly equal to zero? 

### ISSUE THREAD

**Issue by aaronpolhamus** [fchollet/keras#2163]
I'm training a CNN in Keras to differentiate between 196 different classes of automobile's with subtle differences distinguishing them (we're using the [Stanford Cars Dataset](http://ai.stanford.edu/~jkrause/cars/car_dataset.html)). It's a tricky problem, since the number of training examples is small relative to the number of classes (~40 each), so I don't expect great performance. Even if the model is wrong a lot, though, I still expect it to make different guesses based on the data. What is puzzling here is that after training on the following [gist script](https://gist.github.com/aaronpolhamus/034cfe4dd8e2cc26ccb24ad6616bf7b4), both the methods `model.predict` and `model.proba` yield exactly the same thing--a 196-element vector of all zeroes, with the exception of a "1" in the 189th position, e.g.: 

`[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]`

To me this looks like the model is all zeroes, and no updates are being made. This would also explain why the training history from the SGD looks like this: 

`{'loss': [311015.37542976422, 311094.98858055013, 311095.03131139488, 311094.99815815326, 311094.97765225935, 311094.98489685659, 311094.95284872298, 311094.97243369353, 311094.96813605109, 311095.01657662081], 'batch': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'size': [1967, 1967, 1967, 1967, 1967, 1967, 1967, 1967, 1967, 1967]}`

I've seen this behavior across a range of learning rates. Is this simply a case of failing to randomly initialize the model weights? Would we still expect backprop to create changes in weight values? 

I've included 20 post-processed training data examples [here](
https://github.com/fchollet/keras/files/190676/keras_ex_data.zip)

Any pointers very much appreciated.

**Comment by NasenSpray** [comment]
You have to convert Y_train to a one-hot matrix, see [`keras.utils.np_utils.to_categorical`](https://github.com/fchollet/keras/blob/master/keras/utils/np_utils.py#L8-L18)

**Comment by aaronpolhamus** [comment]
@NasenSpray: It looks like this is going to fix my issue. The one thing I unsure about is what seems to be an all-zeros dummy column. E.g., if I have classes 1 - 196, the operation creates a n x 197 dense matrix where the first column is all zeroes. If this necessary for categorical cross entropy?

**Comment by aaronpolhamus** [comment]
ah, ok, figured it out. to_categorical assumes that the first class is coded as a 0. adjusting my labels accordingly solved this problem, either with: 

`Y_train = loadtxt(model_control.y_train_file, delimiter=',', dtype = int)`
`Y_train = Y_train - 1` 
`Y_train = to_categorical(Y_train)`

or 

`Y_train = loadtxt(model_control.y_train_file, delimiter=',', dtype = int)`
`Y_train = to_categorical(Y_train)`
`Y_train = np.delete(Y_train, 0, axis = 1)`

**Comment by aaronpolhamus** [comment]
All sorted, thanks again for the help.
